src/main/java/org/sasanlabs/service/vulnerability/ssrf/SSRFVulnerability.java

```diff
@@ -1,52 +1,221 @@
+                    HttpStatus.OK);
+        }
+    }
+
+    private ResponseEntity<GenericVulnerabilityResponseBean<StringBuilder>>
+            getGenericVulnerabilityResponseWhenFilters(@RequestParam(FILE_URL) String url)
+                    throws IOException {
+        if (!url.startsWith(FILE_PROTOCOL) && !Pattern.matches(LOCALHOST_PATTERN, url)) {
+            return getGenericVulnerabilityResponseWhenURL(url);
+        } else {
+            return new ResponseEntity<>(
+                    new GenericVulnerabilityResponseBean<>(
+                            url.startsWith(FILE_PROTOCOL)
+                                    ? new StringBuilder("file:// protocol is not allowed")
+                                    : new StringBuilder("localhost is not allowed"),
+                            false),
+                    HttpStatus.OK);
+        }
+    }
 
     @AttackVector(
             vulnerabilityExposed = VulnerabilityType.SIMPLE_SSRF,
-            description = "IMAGE_URL_PASSED_TO_REQUEST")
+            description = "FILE_URL_WITHOUT_CHECK",
+            payload = "SSRF_PAYLOAD_LEVEL_1")
     @VulnerableAppRequestMapping(value = LevelConstants.LEVEL_1, htmlTemplate = "LEVEL_1/SSRF")
-    public ResponseEntity<GenericVulnerabilityResponseBean<byte[]>> getVulnerablePayloadLevel1(
-            @RequestParam(IMAGE_URL) String urlImage) {
-        try {
-            URL u = new URL(urlImage);
-            URLConnection urlConnection = u.openConnection();
-            byte[] bytes;
-            try (InputStream in = urlConnection.getInputStream()) {
-                bytes = StreamUtils.copyToByteArray(urlConnection.getInputStream());
+    public ResponseEntity<GenericVulnerabilityResponseBean<StringBuilder>>
+            getVulnerablePayloadLevel1(@RequestParam(FILE_URL) String url)
+                    throws IOException, URISyntaxException {
+        if (isUrlValid(url)) {
+            return getGenericVulnerabilityResponseWhenURL(url);
+        } else {
+            return getGenericVulnerabilityResponseWhenFileName(url);
+        }
+    }
+
+    @AttackVector(
+            vulnerabilityExposed = VulnerabilityType.SSRF_WITH_INPUT_FILTERS,
+            description = "FILE_URL_IF_NOT_FILE_PROTOCOL",
+            payload = "SSRF_PAYLOAD_LEVEL_2")
+    @VulnerableAppRequestMapping(value = LevelConstants.LEVEL_2, htmlTemplate = "LEVEL_1/SSRF")
+    public ResponseEntity<GenericVulnerabilityResponseBean<StringBuilder>>
+            getVulnerablePayloadLevel2(@RequestParam(FILE_URL) String url)
+                    throws IOException, URISyntaxException {
+        if (isUrlValid(url)) {
+            if (!url.startsWith(FILE_PROTOCOL)) {
+                return getGenericVulnerabilityResponseWhenURL(url);
+            } else {
+                return new ResponseEntity<>(
+                        new GenericVulnerabilityResponseBean<>(
+                                new StringBuilder("file:// protocol is not allowed"), false),
```

Comment:
I would suggest, creating a single object as private static final and reuse it to have smaller footprint.